### PR TITLE
[inductor] Improve overlap collective prefetch priority

### PR DESCRIPTION
**Pull request by armijo1122**
## Summary

Improve overlap scheduling for compute/collective reordering by prioritizing collective prefetch candidates that can consume more of the currently available overlap window.

This changes `_schedule_collectives_for_overlap` to:

- skip candidates whose PG has no remaining overlap window
- count `max_coll_distance` over plausible candidates instead of raw unscheduled collective entries
- prefer candidates with larger `min(available_compute_window, collective_exposed_time)`
- preserve the same overlap priority when `prioritize_bucketing_during_scheduling` groups candidates by bucket key

## Motivation

The previous scheduler primarily ordered candidates by semantic priority, compute domination, and original graph order. When multiple collectives had the same semantic priority, this could spend a limited compute overlap window on a small/short collective while leaving a longer collective exposed.

This is suboptimal under memory or in-flight limits. If only one collective can be prefetched before a compute node, the scheduler should prefer the collective that can hide more latency under that compute window, not simply the earlier one in graph order.

There was also a subtle interaction with bucket grouping: even after adding a better candidate priority, `prioritize_bucketing_during_scheduling=True` could regroup and sort candidates by domination/original order, effectively discarding the overlap-window priority in the default path.

## Example / Experimental Observation

The added unit test constructs this case:

- one `mm` provides an 8ms compute overlap window
- `ag_small` has estimated collective time 1ms
- `ag_large` has estimated collective time 8ms
- in-flight memory is limited so only one 128B all_gather can be prefetched
- bucket latency reduction is disabled to isolate scheduling behavior

Old behavior:

| Candidate chosen for overlap | Remaining exposed time |
|---|---|
| `ag_small` is hidden by `mm` | `ag_large` remains exposed, total exposed time = 8ms |

New behavior:

| Candidate chosen for overlap | Remaining exposed time |
|---|---|
| `ag_large` is hidden by `mm` | only `ag_small` remains exposed, total exposed time = 1ms |

This demonstrates that the new priority reduces exposed communication time in a constrained prefetch window.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — COMMENTED** [review]
@armijo1122 would you mind giving a bit of context on how you came across this and if you saw any perf improvements in a model ?

**Comment by armijo1122** [comment]
@eellison I added a simplified standalone benchmark below to try. This is intentionally much smaller than a production model: it is a dense MLP tower rather than a full transformer/training stack. The goal is only to exercise a real end-to-end FSDP2 + DTensor TP path.

On my local H20 setup, the cleaner run using only idle GPUs showed roughly 1-2% improvement. I am not sure whether other model structures or communication/compute ratios would show larger gains.

Standalone script:

```python
# fsdp2_tp_mlp_overlap.py
import argparse
import os
import socket
import statistics
import time

import torch
import torch.distributed as dist
import torch.nn as nn
import torch.nn.functional as F
import torch.multiprocessing as mp
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import MixedPrecisionPolicy, fully_shard
from torch.distributed.tensor import DTensor, Replicate, Shard
from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module


class GatedMLPBlock(nn.Module):
    def __init__(self, hidden, expansion):
        super().__init__()
        self.norm = nn.LayerNorm(hidden)
        self.gate = nn.Linear(hidden, expansion, bias=False)
        self.value = nn.Linear(hidden, expansion, bias=False)
        self.proj = nn.Linear(expansion, hidden, bias=False)

    def forward(self, x):
        h = self.norm(x)
        h = F.silu(self.gate(h)) * self.value(h)
        return x + self.proj(h)


class HeadThenMLP(nn.Module):
    def __init__(self, hidden, expansion, small_head_dim, large_head_dim):
        super().__init__()
        self.small_head = nn.Linear(hidden, small_head_dim, bias=False)
        self.large_head = nn.Linear(hidden, large_head_dim, bias=False)
        self.block = GatedMLPBlock(hidden, expansion)

    def forward(self, x):
        small_shard = self.small_head(x)
        large_shard = self.large_head(x)
        h = self.block(x)
        small = small_shard.redistribute([Replicate()], async_op=True).to_local()
        large = large_shard.redistribute([Replicate()], async_op=True).to_local()
        return h, small.mean() + large.mean()


class DenseTower(nn.Module):
    def __init__(self, layers, hidden, expansion, small_head_dim, large_head_dim, aux_heads):
        super().__init__()
        assert layers > aux_heads
        self.input_proj = nn.Linear(hidden, hidden, bias=False)
        self.first = GatedMLPBlock(hidden, expansion)
        self.stages = nn.ModuleList(
            HeadThenMLP(hidden, expansion, small_head_dim, large_head_dim)
            for _ in range(aux_heads)
        )
        self.tail = nn.ModuleList(
            GatedMLPBlock(hidden, expansion) for _ in range(layers - aux_heads - 1)
        )
        self.norm = nn.LayerNorm(hidden)
        self.score = nn.Linear(hidden, 1, bias=False)

    def forward(self, x):
        h = self.first(self.input_proj(x))
        aux = h.mean()
        for stage in self.stages:
            h, stage_aux = stage(h)
            aux = aux + stage_aux
        for block in self.tail:
            h = block(h)
        return self.score(self.norm(h)).mean() + aux


def add_tp(model, tp_mesh):
    plan = {}

    def add_block(prefix):
        plan[f"{prefix}.gate"] = ColwiseParallel()
        plan[f"{prefix}.value"] = ColwiseParallel()
        plan[f"{prefix}.proj"] = RowwiseParallel()

    add_block("first")
    for i in range(len(model.stages)):
        add_block(f"stages.{i}.block")
        plan[f"stages.{i}.small_head"] = ColwiseParallel(
            output_layouts=Shard(-1), use_local_output=False
        )
        plan[f"stages.{i}.large_head"] = ColwiseParallel(
            output_layouts=Shard(-1), use_local_output=False
        )
    for i in range(len(model.tail)):
        add_block(f"tail.{i}")
    return parallelize_module(model, tp_mesh, plan)


def shard_fn(param):
    if isinstance(param, DTensor):
        for placement in param.placements:
            if isinstance(placement, Shard):
                return Shard(param.ndim - 1 - placement.dim)
    return Shard(0)


def add_fsdp2(model, dp_mesh, dtype, prefetch=1):
    mp_policy = MixedPrecisionPolicy(
        param_dtype=dtype,
        reduce_dtype=torch.float32,
        cast_forward_inputs=False,
    )
    kwargs = dict(
        mesh=dp_mesh,
        mp_policy=mp_policy,
        shard_placement_fn=shard_fn,
        reshard_after_forward=True,
    )
    modules = [model.input_proj, model.first, *model.stages, *model.tail, model.norm, model.score]
    for module in modules:
        fully_shard(module, **kwargs)
    fully_shard(model, **kwargs)
    for i, module in enumerate(modules):
        to_prefetch = modules[i + 1 : i + 1 + prefetch]
        if to_prefetch:
            module.set_modules_to_forward_prefetch(to_prefetch)


def free_port():
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind(("127.0.0.1", 0))
    port = sock.getsockname()[1]
    sock.close()
    return port


def worker(local_rank, args, port):
    os.environ.update(
        RANK=str(local_rank),
        LOCAL_RANK=str(local_rank),
        WORLD_SIZE=str(args.world_size),
        MASTER_ADDR="127.0.0.1",
        MASTER_PORT=str(port),
        USE_LIBUV="0",
    )
    torch.cuda.set_device(local_rank)
    dist.init_process_group("nccl", init_method="env://")

    try:
        dtype = torch.bfloat16
        mesh = init_device_mesh("cuda", (args.dp_size, args.tp_size), mesh_dim_names=("dp", "tp"))
        model = DenseTower(
            args.layers,
            args.hidden,
            args.expansion,
            args.small_head_dim,
            args.large_head_dim,
            args.aux_heads,
        ).cuda().to(dtype)
        model.eval()
        for p in model.parameters():
            p.requires_grad_(False)

        model = add_tp(model, mesh["tp"])
        add_fsdp2(model, mesh["dp"], dtype)
        x = torch.randn(args.batch, args.rows, args.hidden, device="cuda", dtype=dtype)

        patches = {
            "aten_distributed_optimizations.enable_overlap_scheduling": True,
            "aten_distributed_optimizations.collective_bucketing": False,
            "aten_distributed_optimizations.insert_overlap_deps": True,
            "aten_distributed_optimizations.collective_estimator": "analytical",
            "aten_distributed_optimizations.compute_estimator": "analytical",
            "aten_distributed_optimizations.compute_overlap_multipler": 4.0,
            "aten_distributed_optimizations.max_compute_pre_fetch": 200,
            "aten_distributed_optimizations.max_coll_distance": 200,
            "aten_distributed_optimizations.max_in_flight_gb": args.max_in_flight_gb,
            "aten_distributed_optimizations.max_memory_increase_gb": 20.0,
            "aten_distributed_optimizations.max_memory_increase_ratio": 1.0,
            "aten_distributed_optimizations.enable_low_contention_collectives": False,
            "reorder_for_locality": False,
            "triton.native_matmul": False,
            "reorder_for_compute_comm_overlap_passes": [],
            "compile_threads": 1,
            "force_disable_caches": True,
        }
        torch._dynamo.reset()
        torch.cuda.empty_cache()
        t0 = time.perf_counter()
        with torch._inductor.config.patch(patches), torch.no_grad():
            compiled = torch.compile(model, dynamic=False)
            compiled(x)
        torch.cuda.synchronize()
        compile_s = time.perf_counter() - t0
        dist.barrier()

        with torch.no_grad():
            for _ in range(args.warmup):
                compiled(x)
            torch.cuda.synchronize()
            times = []
            for _ in range(args.repeat):
                start = torch.cuda.Event(enable_timing=True)
                end = torch.cuda.Event(enable_timing=True)
                start.record()
                compiled(x)
                end.record()
                end.synchronize()
                times.append(start.elapsed_time(end))
        gathered = [None for _ in range(args.world_size)]
        dist.all_gather_object(gathered, times)
        if local_rank == 0:
            flat = [v for rank_times in gathered for v in rank_times]
            print({
                "compile_s": compile_s,
                "median_ms": statistics.median(flat),
                "p90_ms": sorted(flat)[int(0.9 * (len(flat) - 1))],
                "per_rank_median_ms": [statistics.median(v) for v in gathered],
            })
    finally:
        dist.barrier()
        dist.destroy_process_group()


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--world-size", type=int, default=torch.cuda.device_count())
    parser.add_argument("--dp-size", type=int, default=4)
    parser.add_argument("--tp-size", type=int, default=2)
    parser.add_argument("--batch", type=int, default=1)
    parser.add_argument("--rows", type=int, default=512)
    parser.add_argument("--hidden", type=int, default=2048)
    parser.add_argument("--expansion", type=int, default=8192)
    parser.add_argument("--small-head-dim", type=int, default=4096)
    parser.add_argument("--large-head-dim", type=int, default=65536)
    parser.add_argument("--layers", type=int, default=4)
    parser.add_argument("--aux-heads", type=int, default=2)
    parser.add_argument("--warmup", type=int, default=5)
    parser.add_argument("--repeat", type=int, default=20)
    parser.add_argument("--max-in-flight-gb", type=float, default=0.10223616)
    args = parser.parse_args()
    assert args.world_size == args.dp_size * args.tp_size
    mp.spawn(worker, args=(args, free_port()), nprocs=args.world_size, join=True)


if __name__ == "__main__":
    main()
```

On a cleaner 6-GPU run where I avoided two GPUs with stale driver-side contexts, I used this command instead:

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5 \
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libatomic.so.1 \
TORCHINDUCTOR_CACHE_DIR=.torchinductor_fsdp2_tp_mlp_cache \
python fsdp2_tp_mlp_overlap.py \
  --world-size 6 --dp-size 2 --tp-size 3 \
  --layers 4 --aux-heads 2 \
  --rows 512 --hidden 3072 --expansion 12288 \
  --small-head-dim 6144 --large-head-dim 49152 \
  --warmup 5 --repeat 20 \
  --max-in-flight-gb 0.07
```

For that cleaner 6-GPU setup, I saw two repeat=20 runs with about +0.9% and +2.1% improvement, or roughly +1.5% on average. So I would describe the benefit in this simplified FSDP2+TP MLP-tower scenario as small. This is mostly a sanity/end-to-end validation that the change can help a real hybrid distributed graph; I do not know yet whether other workloads with different communication/compute ratios would show a larger gain.

**Review by eellison — APPROVED** [review]
Looks good to me. cc @IvanKobzarev

**Comment by armijo1122** [comment]
@eellison Could you help trigger CI, I want to see if these changes break any existing tests. Thanks! cc @jansel

**Review by jansel — APPROVED** [review]

**Comment by armijo1122** [comment]
Some pypi conflicts occurred in CI, could you help skip or check if it's unstable or not. Thanks! cc @jansel @eellison

**Comment by armijo1122** [comment]
Still have some pip conflicts, are they related to my changes? Maybe it's not.
`The conflict is caused by:
    torchcomms 0.2.0.dev20260402+cu129 depends on torch==2.12.0.dev20260401
    torchcomms 0.2.0.dev20260401+cu129 depends on torch==2.12.0.dev20260331` cc @jansel @eellison

**Comment by armijo1122** [comment]
By the way, I remember that FSDP hooks are traceable in previous pytorch versions, but in latest main branch FSDP hooks cannot be traced. I am trying to understand the motivation behind this change. Was the traceable hook path removed mainly because of correctness concerns, implementation complexity, compile-time overhead, compiled-autograd limitations, or some other issue?

**Comment by jansel** [comment]
@claude are those CI failures and pip conflicts related to this PR?
